tests/unit/s2n_self_talk_psk_test.c

```diff
@@ -0,0 +1,340 @@
+    GUARD_AS_RESULT(s2n_psk_new_identity(shared_psk, test_shared_identity, sizeof(test_shared_identity)));
+    GUARD_AS_RESULT(s2n_psk_new_secret(shared_psk, test_shared_secret, sizeof(test_shared_secret)));
+    shared_psk->hmac_alg = TEST_PSK_HMAC;
+
+    return S2N_RESULT_OK;
+}
+
+static s2n_result setup_server_psks(struct s2n_connection *server_conn)
+{
+    ENSURE_REF(server_conn);
+
+    /* Setup shared PSK for server */
+    struct s2n_psk *shared_psk = NULL;
+    GUARD_RESULT(s2n_array_pushback(&server_conn->psk_params.psk_list, (void**) &shared_psk));
+    GUARD_AS_RESULT(s2n_psk_init(shared_psk, S2N_PSK_TYPE_EXTERNAL));
+    GUARD_AS_RESULT(s2n_psk_new_identity(shared_psk, test_shared_identity, sizeof(test_shared_identity)));
+    GUARD_AS_RESULT(s2n_psk_new_secret(shared_psk, test_shared_secret, sizeof(test_shared_secret)));
+    shared_psk->hmac_alg = TEST_PSK_HMAC;
+
+    /* Setup other server PSK */
+    uint8_t other_server_data[] = "other server data";
+    struct s2n_psk *other_server_psk = NULL;
+    GUARD_RESULT(s2n_array_pushback(&server_conn->psk_params.psk_list, (void**) &other_server_psk));
+    GUARD_AS_RESULT(s2n_psk_init(other_server_psk, S2N_PSK_TYPE_EXTERNAL));
+    GUARD_AS_RESULT(s2n_psk_new_identity(other_server_psk, other_server_data, sizeof(other_server_data)));
+    GUARD_AS_RESULT(s2n_psk_new_secret(other_server_psk, other_server_data, sizeof(other_server_data)));
+    other_server_psk->hmac_alg = S2N_HMAC_SHA384;
+
+    return S2N_RESULT_OK;
+}
+
+static int s2n_test_select_psk_identity_callback(struct s2n_connection *conn, struct s2n_psk_identity *identities,
+                                                 size_t identities_length, uint16_t *chosen_wire_index)
+{
+    notnull_check(conn);
+    notnull_check(identities);
+    notnull_check(chosen_wire_index);
+
+    struct s2n_array *known_psks = &conn->psk_params.psk_list;
+
+    for (size_t i = 0; i < identities_length; i++) {
+        struct s2n_blob wire_identity = { 0 };
+        GUARD(s2n_blob_init(&wire_identity, identities[i].data, identities[i].length));
+
+        struct s2n_psk *local_match = NULL;
+        GUARD_AS_POSIX(s2n_match_psk_identity(known_psks, &wire_identity, &local_match));
+
+        if (local_match != NULL) {
+            *chosen_wire_index = i;
+            return S2N_SUCCESS;
+        }
+    }
+
+    return S2N_FAILURE;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    /* Happy Case: PSK without callback function */ 
+    {
+        /* Setup connections */
+        struct s2n_connection *client_conn = NULL;
+        struct s2n_connection *server_conn = NULL;
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+
+        /* Setup config */
+        struct s2n_cert_chain_and_key *chain_and_key = NULL;
+        EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key,
```

Comment:
Why are we loading a cert chain? I thought we weren't supposed to be doing that.

Reply:
We would need this for fallback, if PSK is not chosen we fallback to a full handshake.

Reply:
Let's validate at least one handshake without certificates set (failure and success), because that will be a customer use case.